src/devices.jl

```diff
@@ -109,3 +109,59 @@ macro threaded(device, expr)
         end
     end)
 end
+
+"""
+    @time device expr
+
+Device-flexible `@time`.
+
+Lowers to
+```julia
+@time expr
+```
+for CPU devices and
+```julia
+CUDA.@time expr
+```
+for CUDA devices.
+"""
+macro time(device, expr)
+    return esc(quote
+        let
+            if $device isa ClimaComms.CUDADevice
+                CUDA.@time $(expr)
+            else
+                @assert $device isa ClimaComms.AbstractDevice
+                @time $(expr)
+            end
+        end
+    end)
```

Comment:
```suggestion
    return quote
        let
            if $(esc(device)) isa ClimaComms.CUDADevice
                CUDA.@time $(esc(expr))
            else
                @assert $(esc(device)) isa ClimaComms.AbstractDevice
                Base.@time $(esc(expr))
            end
        end
    end
```

Reply:
I think should work

Reply:
not sure you need the `let`?

Reply:
Looks good, I'm trying to write some hygiene tests now, I'll update the `@threaded` code first, then update this PR.

Reply:
I'm running into an issue here (when updating the `@threaded` part):

`test/hygiene.jl`:
```julia
import ClimaComms: @threaded, CPUMultiThreaded, CPUSingleThreaded
dev = CPUSingleThreaded()
@threaded dev for i in 1:3
    1
end
```
error:
```julia
julia> using Revise; include("test/hygiene.jl")
ERROR: LoadError: ArgumentError: @threads requires a `for` loop expression
Stacktrace:
 [1] var"@threads"(__source__::LineNumberNode, __module__::Module, args::Vararg{Any})
   @ Base.Threads ./threadingconstructs.jl:226
 [2] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [3] top-level scope
   @ REPL[2]:1
in expression starting at /dev/CliMA/ClimaComms.jl/src/devices.jl:103
in expression starting at /dev/CliMA/ClimaComms.jl/test/hygiene.jl:3
```
where
```julia
macro threaded(device, expr)
    return quote
            if $(esc(device)) isa CPUMultiThreaded
                Threads.@threads $(esc(expr))
            else
                @assert $(esc(device)) isa CPUSingleThreaded
                $(esc(expr))
            end
        end
end
```